### PR TITLE
feat: introduces OSSM plugin for KfDef

### PR DESCRIPTION
**Pull request by bartoszmajsak**
### :warning: Important notes

The operator bundle has not been officially released yet. If you wish to test it, please use for the time being:

```sh
operator-sdk run bundle quay.io/maistra-dev/opendatahub-operator-bundle:v0.0.10-plugin --namespace openshift-operators --timeout 5m0s
```

It will need https://github.com/opendatahub-io/odh-manifests/pull/932 to be merged alongside to use `service-mesh-integration` branch.

## Description

This PR introduces a `KfDef` plugin aimed at enhancing the configuration and deployment functionalities of OpenShift Service Mesh. The development was inspired by analysis of how other platforms like GCP and AWS manage integrations. We've designed these improvements to be easily integrated into the upcoming version of the opendatahub-operator, while minimizing changes to the existing codebase. All new features are contained within dedicated packages. 

The plugin addresses several limitations we encountered when using the conventional manifest and init-job solution, specifically:

- **Autoconfiguration**: Simplifies the process of obtaining cluster-specific data for auto-configuring authorization settings.
- **Reversible Changes**: Enables flexible and easy reversion of applied configurations.
    - Utilizes owner references wherever applicable.
    - Enables custom cleanup logic hooks implemented in Golang, such as undoing patches.
- **Ease of Integration**: Allows a straightforward addition to the `KfDef` file. An example is provided below.

This PR includes #501 fix.

Manifests used for the plugin can be found here https://github.com/maistra/odh-manifests/tree/ossm_plugin_templates. PR https://github.com/opendatahub-io/odh-manifests/pull/932 

### `KfDef` example

```yaml
apiVersion: kfdef.apps.kubeflow.org/v1
kind: KfDef
metadata:
  name: odh-mesh
spec:
  plugins:
    - kind: KfOssmPlugin
      spec:
        mesh:
          name: minimal
          namespace: istio-system
          installationMode: minimal # (1)
          certificate:
            generate: true # (2)
        auth:
          name: authorino
          namespace: auth-provider
          authorino:
            label: authorino/topic=odh
  applications:
    - kustomizeConfig:
        repoRef:
          name: manifests
          path: odh-common
      name: odh-common
    - kustomizeConfig:
        overlays:
          - dev # (3)
        repoRef:
          name: manifests
          path: odh-dashboard
      name: odh-dashboard
    - kustomizeConfig:
        repoRef:
          name: manifests
          path: odh-notebook-controller
      name: odh-notebook-controller
    - kustomizeConfig:
        repoRef:
          name: manifests
          path: odh-project-controller
      name: odh-project-controller
    - kustomizeConfig:
        repoRef:
          name: manifests
          path: notebook-images
      name: notebook-images
  repos:
    - name: manifests
      uri: https://github.com/maistra/odh-manifests/tarball/kf_ossm_plugin
  version: kf_ossm_plugin

```

`(1)` - `installationMode`: currently it's either `minimal` which deploys our own, very basic control plane, or assumes there's one already which we should use (`installationMode: pre-installed`)

`(2)` - This will generate a self-signed certificate used by Mesh internally. This will need further enhancements.

`(3)` - notice there's no need to explicitly add `service-mesh` overlay. If an application has one defined it will be detected and enabled automatically

## How Has This Been Tested?
- manual end-to-end testing
- automated integration tests
  - new target `test-ossm` has been provided

## Merge criteria:

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: .github/workflows/sync_dashboard_crds.yaml
Comment (bartoszmajsak) [inline .github/workflows/sync_dashboard_crds.yaml:30]:
```suggestion
```

File: bundle/manifests/opendatahub-operator.clusterserviceversion.yaml
Comment (bartoszmajsak) [inline bundle/manifests/opendatahub-operator.clusterserviceversion.yaml:107]:
```suggestion
                "name": "data-science-pipelines-operator"
                 },
                 {
                   "kustomizeConfig": {
                     "repoRef": {
                      "name": "manifests",
                      "path": "trustyai-service"
                    }
                  },
                  "name": "trustyai"
```

File: config/manifests/bases/opendatahub-operator.clusterserviceversion.yaml
Comment (bartoszmajsak) [inline config/manifests/bases/opendatahub-operator.clusterserviceversion.yaml:1]: Some weird merge side-effect. I will investigate.

File: bundle/manifests/opendatahub-operator.clusterserviceversion.yaml
Comment (bartoszmajsak) [inline bundle/manifests/opendatahub-operator.clusterserviceversion.yaml:1]: Some weird merge side-effect. I will investigate.

File: controllers/secretgenerator/secret.go
Comment (bartoszmajsak) [inline controllers/secretgenerator/secret.go:1]: Moved to `pkg/secret`

File: pkg/secret/secret.go
Comment (bartoszmajsak) [inline pkg/secret/secret.go:1]: Somewhat git lost track of move here. I will try to fix it.

File: apis/kfdef.apps.kubeflow.org/v1/kfdef_types.go
Comment (bartoszmajsak) [inline apis/kfdef.apps.kubeflow.org/v1/kfdef_types.go:94]: Without this option we are not able to use custom `struct`s for configuration.

File: controllers/kfdef.apps.kubeflow.org/status.go
Comment (bartoszmajsak) [inline controllers/kfdef.apps.kubeflow.org/status.go:1]: Provided as #501

File: pkg/kfapp/coordinator/coordinator.go
Comment (bartoszmajsak) [inline pkg/kfapp/coordinator/coordinator.go:1]: As the plugin being defined as the platform is not participating in `Apply` and `Delete` phases by default, we need to hook into those in order to configure Service Mesh. GCP integration is done similarly.

File: pkg/kfapp/ossm/feature/raw_resources.go
Comment (bartoszmajsak) [inline pkg/kfapp/ossm/feature/raw_resources.go:1]: This is a copy of k8sresources.go, but I wanted to keep it isolated from the main codebase.

File: pkg/kfapp/ossm/ossm_installer.go
Comment (bartoszmajsak) [inline pkg/kfapp/ossm/ossm_installer.go:1]: This is the central place of kf plugin integration.

Comment (bartoszmajsak) [inline pkg/kfapp/ossm/ossm_installer.go:1]: This is the central place of kf plugin integration.

File: pkg/kfapp/coordinator/coordinator.go
Comment (bartoszmajsak) [inline pkg/kfapp/coordinator/coordinator.go:1]: As the plugin being defined as the platform is not participating in Apply and Delete phases by default, we need to hook into those in order to configure Service Mesh. GCP integration is done similarly.

File: pkg/secret/secret.go
Comment (bartoszmajsak) [inline pkg/secret/secret.go:1]: Not quite sure why it's shown as a new file, as it was first moved and then edited, but in order to see actual changes here's the file diff https://github.com/opendatahub-io/opendatahub-operator/blob/09f889f3a54ad9898cc1c52f2994f948e32bcaf3/pkg/secret/secret.go

File: pkg/kfapp/ossm/OWNERS
Comment (cam-garrison) [inline pkg/kfapp/ossm/OWNERS:2]:
```suggestion
  - bartoszmajsak
  - cam-garrison
```

File: pkg/kfconfig/ossmplugin/OWNERS
Comment (cam-garrison) [inline pkg/kfconfig/ossmplugin/OWNERS:2]:
```suggestion
  - bartoszmajsak
  - cam-garrison
```

**Review by cam-garrison — APPROVED** [review]
Looks great - nice fixes in recent commits. Have also tested this E2E in the past with success using the `Feature` implementation.

**Comment by zdtsw** [comment]
sorry for late chime in @bartoszmajsak : 
we are planning to sunset v1 operator in a couple of months. 
not sure we want to add more new features to it.

**Comment by bartoszmajsak** [comment]
> sorry for late chime in @bartoszmajsak : we are planning to sunset v1 operator in a couple of months. not sure we want to add more new features to it.

@zdtsw that's a valid concern. However, the context here is broader than just sunsetting v1. For ODH users still on v1 who require service mesh, the suggested approach offers clear advantages over our current `init-job` method, as seen in the `service-mesh-integration` branch of the ODH Manifests repo.

Based on the [comment](https://github.com/opendatahub-io/opendatahub-community/issues/97#issuecomment-1689971998) from @LaVLaS I understood we need both. It would be great if he could provide more clarity on this matter.

**Comment by bartoszmajsak** [comment]
> > > sorry for late chime in @bartoszmajsak : we are planning to sunset v1 operator in a couple of months. not sure we want to add more new features to it.
> > 
> > 
> > @zdtsw that's a valid concern. However, the context here is broader than just sunsetting v1. For ODH users still on v1 who require service mesh, the suggested approach offers clear advantages over our current `init-job` method, as seen in the `service-mesh-integration` branch of the ODH Manifests repo.
> > Based on the [comment](https://github.com/opendatahub-io/opendatahub-community/issues/97#issuecomment-1689971998) from @LaVLaS I understood we need both. It would be great if he could provide more clarity on this matter.
> 
> ok, I see. I was not aware of this issue. but yeah, lets wait for more inputs.
> 
> btw, for the manifests, we are doing the transition to move away from `odh-manifests` repo, see [opendatahub-io/architecture-decision-records#16](https://github.com/opendatahub-io/architecture-decision-records/pull/16). from [linked converstation](https://github.com/opendatahub-io/opendatahub-community/issues/97#issuecomment-1689971998) seems you are still updating with the old approach ?

I am aware of this (found out yesterday by accident ;)) and we are already working on the split on our end. But I see not everything is ported yet, e.g. notebook controllers? https://github.com/opendatahub-io/opendatahub-operator/blob/0b2bb98b318580bcae847af6c62303d9170b94b3/get_all_manifests.sh#L22-L32

**Comment by zdtsw** [comment]
> > > > sorry for late chime in @bartoszmajsak : we are planning to sunset v1 operator in a couple of months. not sure we want to add more new features to it.
> > > 
> > > 
> > > @zdtsw that's a valid concern. However, the context here is broader than just sunsetting v1. For ODH users still on v1 who require service mesh, the suggested approach offers clear advantages over our current `init-job` method, as seen in the `service-mesh-integration` branch of the ODH Manifests repo.
> > > Based on the [comment](https://github.com/opendatahub-io/opendatahub-community/issues/97#issuecomment-1689971998) from @LaVLaS I understood we need both. It would be great if he could provide more clarity on this matter.
> > 
> > 
> > ok, I see. I was not aware of this issue. but yeah, lets wait for more inputs.
> > btw, for the manifests, we are doing the transition to move away from `odh-manifests` repo, see [opendatahub-io/architecture-decision-records#16](https://github.com/opendatahub-io/architecture-decision-records/pull/16). from [linked converstation](https://github.com/opendatahub-io/opendatahub-community/issues/97#issuecomment-1689971998) seems you are still updating with the old approach ?
> 
> I am aware of this (found out yesterday by accident ;)) and we are already working on the split on our end. But I see not everything is ported yet, e.g. notebook controllers?
> 
> https://github.com/opendatahub-io/opendatahub-operator/blob/0b2bb98b318580bcae847af6c62303d9170b94b3/get_all_manifests.sh#L22-L32

good to know that!

`notebook-controller` actually is ready , i am waiting for "approval" on https://github.com/opendatahub-io/opendatahub-operator/pull/543  to close it.
but for the `model-mesh` part that will take longer time ( i guess) 

it is up to when servicemesh can finish "split" work.
But no matter in which case, we will need to add it in the `get_all_manifests.sh` either from your repo or from `odh-manifests`

**Comment by bartoszmajsak** [comment]
@zdtsw is this split going to affect v1/KfDef operator as well?

**Comment by zdtsw** [comment]
> @zdtsw is this split going to affect v1/KfDef operator as well?

yes, the PR for v1 is here https://github.com/opendatahub-io/opendatahub-operator/pull/544
similar way to handle it, but different components in v1

**Comment by bartoszmajsak** [comment]
Withdrawing in favor of Operator V2 support #605

**Pull request closed without merging**
